### PR TITLE
Feature/container image from data

### PR DESCRIPTION
**Pull request by disconsented**
Might be a bit domain specific, allows for environments to support token replacements.

Used to allow me to have container images specified via data, close if unwanted

### CODE REVIEW

**Comment by LordRalex** [comment]
It seems like we could just handle this on the executeAsync (calculate the image name) and pass it around, instead of creating a pretty much docker-specific interface that is a bit mis-leading.

**Comment by disconsented** [comment]
This was born out of the need for changing Java versions, where I decided that this would be good to be a bit more generic now.

I thought executeAsync was just for sending commands to an environment .. ?

**Comment by LordRalex** [comment]
It seems like as part of our execution data, we should pass the information we need (including the map) and let the environment handle replacements if needed.

In actuality, your code would also only work once. The first time you run the execute, it does a full replacement of the imageName variable. Next time it runs, it won't have the "variable", so it won't be replacing anything.

**Comment by disconsented** [comment]
There we go, that should have that all sorted.

File: environments/docker/docker.go [hunk 1]
```diff
@@ -84,6 +84,8 @@ func (d *docker) dockerExecuteAsync(steps pufferpanel.ExecutionData) error {
 		return errors.New("docker container already exists")
 	}
 
+	d.ImageName = pufferpanel.ReplaceTokens(d.ImageName, steps.Data)
```
Comment (LordRalex) [inline environments/docker/docker.go:87]: I still don't like the idea of overriding the value, because this would persist.

**Comment by LordRalex** [comment]
Replaced with https://github.com/PufferPanel/PufferPanel/commit/c3bb94b5207d3b62e682db713e96b9805150935c

**Pull request closed without merging**